HLTrigger/Muon/interface/HLTTriMuonIsolation.h

```diff
@@ -0,0 +1,230 @@
+}
+
+HLTTriMuonIsolation::~HLTTriMuonIsolation(){ }
+
+void
+HLTTriMuonIsolation::produce(edm::StreamID sid, edm::Event & iEvent, edm::EventSetup const & iSetup) const
+{
+    std::unique_ptr<reco::CompositeCandidateCollection> Taus        (new reco::CompositeCandidateCollection);
+    std::unique_ptr<reco::CompositeCandidateCollection> SelectedTaus(new reco::CompositeCandidateCollection);
+
+    // Get the L3 muon candidates
+    edm::Handle<reco::RecoChargedCandidateCollection> L3MuCands;
+    iEvent.getByToken(L3MuonsToken_, L3MuCands);
+
+    // Get the L3 muon candidates that passed the filter
+    edm::Handle<trigger::TriggerFilterObjectWithRefs> L3DiMuonsFilterCands;
+    iEvent.getByToken(L3DiMuonsFilterToken_, L3DiMuonsFilterCands);
+    
+    std::vector<reco::RecoChargedCandidateRef> PassedL3Muons;
+    L3DiMuonsFilterCands->getObjects(trigger::TriggerMuon, PassedL3Muons);
+  
+    // Get the Trk + L3 muon candidates (after merging)
+    edm::Handle<reco::RecoChargedCandidateCollection> AllMuCands;
+    iEvent.getByToken(AllMuonsToken_, AllMuCands);
+    
+    // Get iso tracks
+    edm::Handle<reco::TrackCollection> IsoTracks;
+    iEvent.getByToken(IsoTracksToken_, IsoTracks);
+          
+    if (AllMuCands->size() >= 3){
+        // Create the 3-muon candidates
+        // loop over L3/Trk muons and create all combinations
+        for (unsigned int i = 0; i != AllMuCands->size(); ++i){
+            const reco::TrackRef &tk_i = (*AllMuCands)[i].track();
+            for (unsigned int j = i+1; j != AllMuCands->size(); ++j){
+                const reco::TrackRef &tk_j = (*AllMuCands)[j].track();
+                for (unsigned int k = j+1; k != AllMuCands->size(); ++k){
+                    const reco::TrackRef &tk_k = (*AllMuCands)[k].track();
+
+                    int passingPreviousFilter_1 = 0;
+                    int passingPreviousFilter_2 = 0;
+                    int passingPreviousFilter_3 = 0;
+    
+                    for (std::vector<reco::RecoChargedCandidateRef>::const_iterator imu = PassedL3Muons.begin(); imu != PassedL3Muons.end(); ++imu){
+                        reco::TrackRef candTrkRef = (*imu)->get<reco::TrackRef>();
+                        if (reco::deltaR2(tk_i->momentum(), candTrkRef->momentum()) < (0.03*0.03)) passingPreviousFilter_1++;
+                        if (reco::deltaR2(tk_j->momentum(), candTrkRef->momentum()) < (0.03*0.03)) passingPreviousFilter_2++; 
+                        if (reco::deltaR2(tk_k->momentum(), candTrkRef->momentum()) < (0.03*0.03)) passingPreviousFilter_3++;
+                    }                    
+                    // at least two muons must have passed the previous di-muon filter
+                    if (((passingPreviousFilter_1 < 1) & (passingPreviousFilter_2 < 1)) ||
+                        ((passingPreviousFilter_1 < 1) & (passingPreviousFilter_3 < 1)) ||
+                        ((passingPreviousFilter_2 < 1) & (passingPreviousFilter_3 < 1))) continue;
+
```

Comment:
IOW, the above should be caluclated less often in the corresponding loop for i/1, j/2 and k/3.